parquet-encoding/src/main/java/parquet/bytes/CapacityByteArrayOutputStream.java

```diff
@@ -38,66 +42,73 @@
 public class CapacityByteArrayOutputStream extends OutputStream {
   private static final Log LOG = Log.getLog(CapacityByteArrayOutputStream.class);
 
-  private static final int MINIMUM_SLAB_SIZE = 64 * 1024;
-  private static final int EXPONENTIAL_SLAB_SIZE_THRESHOLD = 10;
+  private static final byte[] EMPTY_SLAB = new byte[0];
 
-  private int slabSize;
+  private int initialSize;
+  private final int pageSize;
   private List<byte[]> slabs = new ArrayList<byte[]>();
   private byte[] currentSlab;
-  private int capacity;
+  private int capacity = 0;
   private int currentSlabIndex;
   private int currentSlabPosition;
   private int size;
 
   /**
-   * @param initialSize the initialSize of the buffer (also slab size)
+   * defaults pageSize to 1MB
+   * @param initialSize
+   * @deprecated use {@link CapacityByteArrayOutputStream#CapacityByteArrayOutputStream(int, int)}
    */
+  @Deprecated
   public CapacityByteArrayOutputStream(int initialSize) {
-    Preconditions.checkArgument(initialSize > 0, "initialSize must be > 0");
+    this(initialSize, 1024 * 1024);
+  }
+
+  /**
+   * @param initialSize the initialSize of the buffer (also slab size)
+   * @param pageSize
+   */
+  public CapacityByteArrayOutputStream(int initialSize, int pageSize) {
+    checkArgument(initialSize > 0, "initialSize must be > 0");
+    checkArgument(pageSize > 0, "pageSize must be > 0");
+    this.pageSize = pageSize;
     initSlabs(initialSize);
   }
 
   private void initSlabs(int initialSize) {
     if (Log.DEBUG) LOG.debug(String.format("initial slab of size %d", initialSize));
-    this.slabSize = initialSize;
+    this.initialSize = initialSize;
     this.slabs.clear();
-    this.capacity = initialSize;
-    this.currentSlab = new byte[slabSize];
-    this.slabs.add(currentSlab);
-    this.currentSlabIndex = 0;
+    this.capacity = 0;
+    this.currentSlab = EMPTY_SLAB;
+    this.currentSlabIndex = -1;
     this.currentSlabPosition = 0;
     this.size = 0;
   }
 
+  /**
+   * the new slab is guaranteed to be at least minimumSize
+   * @param minimumSize the size of the data we want to copy in the new slab
+   */
   private void addSlab(int minimumSize) {
     this.currentSlabIndex += 1;
-    if (currentSlabIndex < this.slabs.size()) {
-      // reuse existing slab
-      this.currentSlab = this.slabs.get(currentSlabIndex);
-      if (Log.DEBUG) LOG.debug(String.format("reusing slab of size %d", currentSlab.length));
-      if (currentSlab.length < minimumSize) {
-        if (Log.DEBUG) LOG.debug(String.format("slab size %,d too small for value of size %,d. replacing slab", currentSlab.length, minimumSize));
-        byte[] newSlab = new byte[minimumSize];
-        capacity += minimumSize - currentSlab.length;
-        this.currentSlab = newSlab;
-        this.slabs.set(currentSlabIndex, newSlab);
-      }
+    int nextSlabSize;
+    if (size == 0) {
+      nextSlabSize = initialSize;
+    } else if (size > pageSize / 5) {
```

Comment:
should 5 be configurable too?

we could also make CapacityByteArrayOutputStream abstract or take as an argument a slab size calculator etc. so that we can plug in different behaviors here. what do you think?